### PR TITLE
Bundle arguments to ChainSync.Client and internally refactor the core definition

### PR DESCRIPTION
**Pull request by nfrisby**
While working on PR 808, I feel like the ChainSync client module needs some refactoring.

- It takes way too many arguments, so I've introduced data types that bundle those.
- It has way too many declarations in its `where`-clause, and with the new argument bundles, it's easier than ever before to float those out to their own top-level definitions.
- Moreover, some lines of logic in the message handlers can also be floated out as their own separate logic. This makes the definitions of the higher-level steps easier to read---eg, each one fits on a single "page" now.
- Finally: the whole module's styling is not internally consistent.

### CODE REVIEW

File: ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs
Comment (nfrisby) [inline ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs]: Still binding the type variable, as in the previous commit.

**Comment by nfrisby** [comment]
@dnadales @RenateEilers The last two commits have very ugly diffs.

- The last commit has hundreds of lines of layout changes, due to it being a restyling commit.
- The second-to-last commit moved a lot of code around, since it's floating out existing definitions and folding chunks to create some new definitions.

For the last commit, I recommend using a smart differ, eg

```
$ x=<the restyling commit hash>
$ >before.hs git show ${x}^:ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs
$ >after.hs git show $x:ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs
$ nix-shell -p difftastic --run 'difft before.hs after.hs'
```

For the second-to-last commit, I recommend having the two files open side-by-side in your editor, so you can manually cross-reference the new definitions versus their previous location in the preceding commit as you come across each of them in the commit's diff.

```
$ x=<the rearranging commit hash>
$ >before.hs git show ${x}^:ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs
$ >after.hs git show $x:ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs
$ emacs before.hs after.hs &
$ git show $x
```

If that's untenable, then I could split that commit into smaller pieces, one per new definition. But I'd rather avoid that work, if the editor-based method works for you. Please let me know 👍

**Comment by nfrisby** [comment]
I'm leaving it in Draft until we assess the performance concerns that @jasagredo raised about thunks/laziness/etc.

**Comment by nfrisby** [comment]
I pushed up commits that add a small "benchmark" exe that dumps a lot of measurements. I eyeballed them to convince myself that the time behavior hasn't changed in this PR.

Maybe that benchmark should be split into a different PR before merging this one.

I'm still working on convincing myself about space leaks.

**Comment by nfrisby** [comment]
No indication of a space leak in the tip of this PR; 10x 50x and 100x number of messages didn't change the max residency of the benchmark.

```
$ git diff --stat; git diff --stat --cached; git lg -1; cabal build ChainSync-client-bench; for i in 10 50 100; do echo $i; time cabal run ChainSync-client-bench -- $i $i +RTS -s 2>&1 1>/dev/null | grep -e
 'maximum residency' ; done
8cc6d46ee - (HEAD -> nfrisby/ChainSync-refactor, origin/nfrisby/ChainSync-refactor) Update changelogs for ChainSync.Client refactor (2 days ago) <Nicolas Frisby>
Up to date
10
         118,736 bytes maximum residency (10 sample(s))

real	0m9.557s
user	0m9.440s
sys	0m0.267s
50
         118,848 bytes maximum residency (43 sample(s))

real	0m41.005s
user	0m40.943s
sys	0m0.061s
100
         118,768 bytes maximum residency (51 sample(s))

real	1m23.457s
user	1m23.254s
sys	0m0.152s
```

**Comment by nfrisby** [comment]
I'd like to use `NoThunks` to test this, but it's currently blocked on https://github.com/input-output-hk/cardano-base/pull/449 and so on. (I could do without it, but it'd require a good deal of yak-shaving.)

File: ouroboros-consensus/bench/ChainSync-client-bench/Bench/Consensus/ChainSyncClient/Driver.hs
Comment (RenateEilers) [inline ouroboros-consensus/bench/ChainSync-client-bench/Bench/Consensus/ChainSyncClient/Driver.hs:6]: Just curious, why 'UA'?

File: ouroboros-consensus-diffusion/src/ouroboros-consensus-diffusion/Ouroboros/Consensus/Network/NodeToNode.hs
Comment (jasagredo) [inline ouroboros-consensus-diffusion/src/ouroboros-consensus-diffusion/Ouroboros/Consensus/Network/NodeToNode.hs:586]: I find it confusing that a qualified field inherits the unqualified value of the same unqualified name.

File: ouroboros-consensus/bench/ChainSync-client-bench/Bench/Consensus/ChainSyncClient/Driver.hs
Comment (jasagredo) [inline ouroboros-consensus/bench/ChainSync-client-bench/Bench/Consensus/ChainSyncClient/Driver.hs:6]: Why `UA`?

File: ouroboros-consensus/bench/ChainSync-client-bench/Main.hs
Comment (jasagredo) [inline ouroboros-consensus/bench/ChainSync-client-bench/Main.hs:287]:
```suggestion
        ChainDB.followerClose = pure ()
      
      , ChainDB.followerForward = \case
```

I would prefer commas before names, even if we leave an empty line in between. But this is just an opinion.

File: ouroboros-consensus/changelog.d/20231206_171848_nick.frisby_ChainSync_refactor.md
Comment (jasagredo) [inline ouroboros-consensus/changelog.d/20231206_171848_nick.frisby_ChainSync_refactor.md:22]: I would mention the type names for quick grepping if needed.

File: ouroboros-consensus/changelog.d/20231206_171848_nick.frisby_ChainSync_refactor.md
Comment (jasagredo) [inline ouroboros-consensus/changelog.d/20231206_171848_nick.frisby_ChainSync_refactor.md:26]:
```suggestion
- Also introduced a `SomeHeaderInFutureCheck` that binds the existential type
  variables separately from the actual payload of the `HeaderInFutureCheck`
  record type.
```

Let's take advantage of this being markdown

File: ouroboros-consensus/ouroboros-consensus.cabal
Comment (jasagredo) [inline ouroboros-consensus/ouroboros-consensus.cabal:699]: I think it is worth grouping this in a common stanza for benchmarks.

File: ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs
Comment (jasagredo) [inline ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs:475]: Despite appreciating the attempt to reduce renaming, once we are in the journey of making a breaking change I would rather prefix the fields with `ce` at least.

File: ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs
Comment (jasagredo) [inline ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs:486]: Same

File: ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs
Comment (jasagredo) [inline ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs:25]: I would use this opportunity to link to the relevant external documentation here, like `typed-protocols` and the mini-protocols document as well as the ouroboros-network implementation of the ChainSync miniprotocol.

**Review by jasagredo — COMMENTED** [review]
I can suggest adding this initial `.hlint.yaml` file:

```yaml
- modules:
  - { name: Ouroboros.Consensus.MiniProtocol.ChainSync.Client
    , importStyle: qualified
    , message: "Exports fields with very common names"
    }

- ignore: {}

- warn: {name: Ouroboros.Consensus.MiniProtocol.ChainSync.Client should be imported qualified}
```

Which will only trigger the import warning. It already fails for:
```
ouroboros-consensus\test\consensus-test\Test\Consensus\MiniProtocol\ChainSync\Client.hs
ouroboros-consensus-diffusion\src\ouroboros-consensus-diffusion\Ouroboros\Consensus\Node\ErrorPolicy.hs
ouroboros-consensus-diffusion\src\ouroboros-consensus-diffusion\Ouroboros\Consensus\Node\RethrowPolicy.hs
ouroboros-consensus-diffusion\src\ouroboros-consensus-diffusion\Ouroboros\Consensus\Node\Tracers.hs
ouroboros-consensus-diffusion\test\consensus-test\Test\Consensus\PeerSimulator\Run.hs
ouroboros-consensus-diffusion\test\consensus-test\Test\Consensus\PeerSimulator\Trace.hs
```

I left some comments but I'm approving because I think they are not severe.

**Review by jasagredo — APPROVED** [review]
Approving because what I said above is not severe.

File: ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client/InFutureCheck.hs
Comment (RenateEilers) [inline ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client/InFutureCheck.hs:139]: Why was `max 0` added? Can `onset` occur after `now`?

File: ouroboros-consensus/src/unstable-consensus-testlib/Test/Util/TestBlock.hs
Comment (RenateEilers) [inline ouroboros-consensus/src/unstable-consensus-testlib/Test/Util/TestBlock.hs:832]: Should this be `chain of lenght L`?

File: ouroboros-consensus/bench/ChainSync-client-bench/Bench/Consensus/ChainSyncClient/Driver.hs
Comment (RenateEilers) [reply ouroboros-consensus/bench/ChainSync-client-bench/Bench/Consensus/ChainSyncClient/Driver.hs:6]: I had the same question, and saw later on that it's because of the unboxed arrays in the module (I think). Do we have any guidelines when it comes to naming imports?

Comment (jasagredo) [reply ouroboros-consensus/bench/ChainSync-client-bench/Bench/Consensus/ChainSyncClient/Driver.hs:6]: We don't really have many conventions. The only ones I can think of from the top of my head are:
- `Ledger.Core` is usually `Core` or `LC`
- `Consensus.Util.IOLike` as `IOLike`
- `Data.X.Strict` as `X` (with `X` being a container say `Map` or `Set`).
- `CBOR....` as `CBOR`
- `...AnchoredFragment` as `AF`
- `...AnchoredSeq` as `AS`
- `Data.ByteString` as `BS`
- `Data.ByteString.Lazy` as `BSL`
- `Data.Text` as `Text` or `T`
- `Test.StateMachine.*` as `QSM`
But we are not very consistent with this I think.

File: ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs
Comment (nfrisby) [reply ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs:475]: In the call, we agreed to keep unprefixed names. But we're going to add a hlint rule that required qualified import and limits the names to `CSC` or `CSClient` or `CsClient`.

File: ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs
Comment (nfrisby) [reply ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs:486]: Same

File: ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client/InFutureCheck.hs
Comment (nfrisby) [reply ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client/InFutureCheck.hs]: Thanks for asking! It's trickier than I thought: see https://github.com/input-output-hk/io-sim/issues/129

File: ouroboros-consensus/bench/ChainSync-client-bench/Main.hs
Comment (nfrisby) [reply ouroboros-consensus/bench/ChainSync-client-bench/Main.hs:287]: Done

File: ouroboros-consensus/ouroboros-consensus.cabal
Comment (nfrisby) [reply ouroboros-consensus/ouroboros-consensus.cabal:699]: Done

File: ouroboros-consensus/src/unstable-consensus-testlib/Test/Util/TestBlock.hs
Comment (nfrisby) [reply ouroboros-consensus/src/unstable-consensus-testlib/Test/Util/TestBlock.hs:832]: That's a reasonable expectation! But because a block chain has the Merkle tree property, a chain and the block at its tip are in many ways indistinguishable. In particular, the length of a block is exactly the length of _the_ chain that ends in that block.

So you'll see this language elsewhere; sometimes "chain" or "block" feels more appropriate depending on context.

File: ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs
Comment (nfrisby) [reply ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs:475]: I finally found the HLint manual: https://github.com/ndmitchell/hlint/blob/master/README.md#customizing-the-hints

FYI: we want `explicitOrQualified` instead of just `qualified`.

Comment (nfrisby) [reply ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs:475]: However, it seems to not work :/ I'll open an Issue to investigate.

Comment (nfrisby) [reply ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs:475]: Doh! Our `.gitignore` explicitly forbids `.hlint.yaml`. I don't want to shave this yak in this PR.

Comment (nfrisby) [reply ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs:475]: I opened https://github.com/input-output-hk/ouroboros-consensus/issues/833

File: ouroboros-consensus-diffusion/src/ouroboros-consensus-diffusion/Ouroboros/Consensus/Network/NodeToNode.hs
Comment (nfrisby) [reply ouroboros-consensus-diffusion/src/ouroboros-consensus-diffusion/Ouroboros/Consensus/Network/NodeToNode.hs:586]: Yeahs, it's a weird corner case of `-XNamedFieldPuns`.

File: ouroboros-consensus/changelog.d/20231206_171848_nick.frisby_ChainSync_refactor.md
Comment (nfrisby) [reply ouroboros-consensus/changelog.d/20231206_171848_nick.frisby_ChainSync_refactor.md:22]: Done

**Comment by nfrisby** [comment]
Last step before merging will be me locally patching things to build with `NoThunks` using Joris' most recent CHaP release.

**Comment by nfrisby** [comment]
OK, thanks to Joris's help, I have run the benchmark with `NoThunks`-checking enabled and no thunks were found 🖐️ 

I pushed my commit to `nfrisby/ChainSync-refactor-NoThunks`.

Using that commit, I:

- built the benchmark _without_ the flags
- added the flags to my `cabal.project`
- rebuilt the benchmark, confirming that the relevant packages were rebuilt
- ran the benchmark (with arguments 1 1, since it's very slow now)

I saw no `NoThunks` messages, merely the normal benchmark output.

**Comment by nfrisby** [comment]
The last few force pushes were just be supplicating to the .cabal ahd .hs style checkers.

**Comment by nfrisby** [comment]
And this last force-push was because I forgot again about signing my commits.

**Pull request merged**
